server/api/db/sqldb/db.py

```diff
@@ -1121,47 +1121,57 @@ def _mark_best_iteration_artifact(
         if link_key:
             key = link_key
 
-        # We perform two consecutive SELECT queries and modify the two artifact records that are returned.
-        # Without no_autoflush, SQLAlchemy offloads the transaction data to the DB during the second SELECT query,
-        # which can result in a deadlock due to assumed conflicts between the transactions.
-        # Using no_autoflush ensures that all modifications don't arrive to the DB until the transaction is committed.
-        with session.no_autoflush:
-            # get the best iteration artifact record
-            query = self._query(session, ArtifactV2).filter(
-                ArtifactV2.project == project,
-                ArtifactV2.key == key,
-                ArtifactV2.iteration == link_iteration,
-            )
-            if link_tree:
-                query = query.filter(ArtifactV2.producer_id == link_tree)
-            if uid:
-                query = query.filter(ArtifactV2.uid == uid)
-
-            best_iteration_artifact_record = query.one_or_none()
-            if not best_iteration_artifact_record:
-                raise mlrun.errors.MLRunNotFoundError(
-                    f"Best iteration artifact not found - {project}/{key}:{link_iteration}",
-                )
+        # Lock the artifacts with the same project and key (and producer_id when available) to avoid unexpected
+        # deadlocks and conform to our lock-once-when-starting logic - ML-6869
+        lock_query = self._query(
+            session,
+            ArtifactV2,
+            project=project,
+            key=key,
+        ).with_entities(ArtifactV2.id)
+        if link_tree:
+            lock_query = lock_query.filter(ArtifactV2.producer_id == link_tree)
```

Comment:
Before this change we had 2 queries here that always returned different records (if returning anything). Now we're adding a 3rd query that contains the results of the other 2. So, there's a question - is it better to retrieve just the IDs here and go to the DB twice more for the other queries, or maybe it's better to retrieve the full objects here and perform the processing client-side on the results without querying again?
On one hand, the current approach has less memory and network requirements since only getting IDs here, and retrieving full objects only for pin-point queries with at most 1 result. OTOH, we do 3 queries with the DB, and this first query is not expected to return many results since it only asks for a specific key within a specific tree (and multiple iterations, true). 
I'm raising this here, but I really am not sure what's better... I guess our memory/processing load is higher than query overhead on the DB, so we can keep it like this.

Reply:
since we have been burnt by these memory issues again I prefer keeping it like this as you said. but I agree this is a question